File: go/tracing.go
package aight

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"sync"
)

// SDKVersion must match projects.CURRENT_SDK_VERSIONS["go"] on the backend —
// bump both together on release.
const SDKVersion = "0.1.0"

// thisFile identifies this package's own source file at runtime, so
// resolveCallerFrame can walk past it to find the caller's actual line —
// mirrors python/aight/tracing.py's _resolve_caller_frame.
var thisFile = func() string {
	_, file, _, _ := runtime.Caller(0)
	return file
}()

// sdkDir is this package's own directory. The walk below skips every frame in
// it except a _test.go one, so TracedHTTPClient's transport — a second file of
// ours, sitting between the caller and net/http — is skipped too without
// listing our files by name. The test files are exempt because they are
// callers: Go keeps them in the package directory, and their call sites are
// exactly what should be recorded.
var sdkDir = filepath.Dir(thisFile)

// SnippetContext is how many source lines are grabbed on each side of an
// attributed line for the Workspace's line-review code panel. A fixed small
// window is plenty for "what's on this line" and keeps payloads small — the
// same constant, and the same reasoning, as the Python SDK's _SNIPPET_CONTEXT.
const SnippetContext = 3

// SnippetKey identifies one attributed source line: the (filepath, lineno)
// pair a snippet belongs to.
type SnippetKey struct {
	Filepath string
	Lineno   int
}

// Snippet is the source window captured around one attributed line.
type Snippet struct {
	// Start is the 1-indexed line number Text's first line came from.
	Start int
	Text  string
}

// Snippets maps each attributed line ever seen to its captured source window.
// Push reads it to attach context to each chain frame; an absent entry means
// "not captured" and degrades to the Workspace's existing empty state.
var (
	snippetsMu sync.Mutex
	Snippets   = map[SnippetKey]Snippet{}
)

// snippetFor returns the captured snippet for one line, if there is one.
func snippetFor(filepath string, lineno int) (Snippet, bool) {
	snippetsMu.Lock()
	defer snippetsMu.Unlock()
	s, ok := Snippets[SnippetKey{Filepath: filepath, Lineno: lineno}]
	return s, ok
}

// captureSnippet reads filepath and stores a SnippetContext-line window around
// lineno, once per line. Best-effort: an unreadable file (deployed without
// source, a path from another machine, a stdin frame) is skipped silently — a
// missing snippet degrades to "not captured", it is never worth raising over.
func captureSnippet(filepath string, lineno int) {
	if _, ok := snippetFor(filepath, lineno); ok {
		return
	}
	start := lineno - SnippetContext
	if start < 1 {
		start = 1
	}
	// Read outside the lock: a racing duplicate read is harmless (identical
	// content, last write wins), holding a global mutex across file I/O is not.
	data, err := os.ReadFile(filepath)
	if err != nil {
		return
	}
	// CRLF in, LF out — the panel renders this text, not a terminal.
	lines := strings.Split(strings.ReplaceAll(string(data), "\r\n", "\n"), "\n")
	if start > len(lines) {
		return
	}
	end := lineno + SnippetContext
	if end > len(lines) {
		end = len(lines)
	}
	snippetsMu.Lock()
	Snippets[SnippetKey{Filepath: filepath, Lineno: lineno}] = Snippet{
		Start: start,
		// Split leaves a trailing empty element for a file that ends in a
		// newline, which would join back in as a phantom last line.
		Text: strings.TrimSuffix(strings.Join(lines[start-1:end], "\n"), "\n"),
	}
	snippetsMu.Unlock()
}

// BucketKey identifies one traced (file, line, function, model). The model is
// part of the identity, not a field read off the bucket: a line that calls two
// models stays two buckets, each priced on its own tokens. Merging them summed
// both models' tokens under whichever name was recorded last, and the server
// then repriced every cheap call at the expensive model's rate.
type BucketKey struct {
	Filepath string
	Lineno   int
	Function string
	Model    string
}

// Bucket aggregates calls and tokens for one BucketKey. No cost: only the
// platform prices a call (see TracedLLMCall).
type Bucket struct {
	Calls        int
	InputTokens  int
	OutputTokens int
	// Kept beside InputTokens, never folded into it. The platform sums four
	// rates and freezes the result at receive, so an input count that already
	// contained the cached tokens would bill them twice — once at the full
	// input rate and again at the cache rate — and the overcount could never
	// be corrected.
	CacheReadTokens     int
	CacheCreationTokens int
	// Summed, not averaged: the server divides by Calls when it stores the row.
	LatencyMs float64
	// The last non-empty trace id recorded against this bucket, so an explicit
	// id from the run that produced the spend lands on its row. Empty means the
	// caller never set one, and the server falls back to its time-window match.
	TraceID string
}

// Site returns the key with the model cleared — the "where in the code" half
// of the identity, which is all resolveCallerFrame can determine on its own.
func (k BucketKey) Site() BucketKey {
	k.Model = ""
	return k
}

// CostByLineProcessor aggregates calls and tokens by the (file, line, function)
// that issued them — the attribution half of the picture; the platform
// computes the money half.
type CostByLineProcessor struct {
	mu      sync.Mutex
	Buckets map[BucketKey]*Bucket
}

func newProcessor() *CostByLineProcessor {
	return &CostByLineProcessor{Buckets: make(map[BucketKey]*Bucket)}
}

// CostProcessor is the package-level singleton every TracedLLMCall records into.
var CostProcessor = newProcessor()

func (p *CostByLineProcessor) record(
	key BucketKey, inputTokens, outputTokens, cacheReadTokens, cacheCreationTokens int,
	latencyMs float64, traceID string) {
	p.mu.Lock()
	defer p.mu.Unlock()
	b, ok := p.Buckets[key]
	if !ok {
		b = &Bucket{}
		p.Buckets[key] = b
	}
	b.Calls++
	b.InputTokens += inputTokens
	b.OutputTokens += outputTokens
	b.CacheReadTokens += cacheReadTokens
	b.CacheCreationTokens += cacheCreationTokens
	b.LatencyMs += latencyMs
	// Last non-empty wins, not last-set: a later call that reported no trace id
	// must not blank the id an earlier call on the same line did report.
	if traceID != "" {
		b.TraceID = traceID
	}
}

// Clear resets all recorded buckets — mainly useful between test runs.
func (p *CostByLineProcessor) Clear() {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.Buckets = make(map[BucketKey]*Bucket)
}

// Snapshot returns a copy of the current buckets, safe to range over concurrently.
func (p *CostByLineProcessor) Snapshot() map[BucketKey]Bucket {
	p.mu.Lock()
	defer p.mu.Unlock()
	out := make(map[BucketKey]Bucket, len(p.Buckets))
	for k, b := range p.Buckets {
		out[k] = *b
	}
	return out
}

// Report renders a plain-text calls-by-line report, busiest line first. No
// dollar figure: this SDK holds no prices, so the honest local report is what
// it actually knows — where the calls happened, and how many tokens each line
// sent. The platform prices them (see TracedLLMCall).
func (p *CostByLineProcessor) Report() string {
	snap := p.Snapshot()
	keys := make([]BucketKey, 0, len(snap))
	for k := range snap {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool { return snap[keys[i]].Calls > snap[keys[j]].Calls })

	lines := make([]string, 0, len(keys))
	for _, k := range keys {
		b := snap[k]
		model := ""
		if k.Model != "" {
			model = fmt.Sprintf(" [%s]", k.Model)
		}
		lines = append(lines, fmt.Sprintf("%s:%d (%s)%s = %d call(s), %d in / %d out",
			filepath.Base(k.Filepath), k.Lineno, k.Function, model, b.Calls, b.InputTokens, b.OutputTokens))
	}
	return strings.Join(lines, "\n")
}

func resolveCallerFrame() BucketKey {
	// Skip resolveCallerFrame and TracedLLMCall's own frames (2), then walk
	// past this file itself in case of extra wrapper frames.
	pc := make([]uintptr, 32)
	n := runtime.Callers(2, pc)
	frames := runtime.CallersFrames(pc[:n])
	for {
		frame, more := frames.Next()
		// Skip this SDK's own frames, and net/http's — a call routed through
		// TracedHTTPClient would otherwise attribute to http.Client.Do rather
		// than to the line that issued it. The skip is on the package path in
		// the function name, so a user package that happens to be named "http"
		// still resolves to its own frame.
		ownSource := filepath.Dir(frame.File) == sdkDir && !strings.HasSuffix(frame.File, "_test.go")
		if !ownSource && !strings.HasPrefix(frame.Function, "net/http.") {
			fn := frame.Function
			if i := strings.LastIndex(fn, "."); i != -1 {
				fn = fn[i+1:] // drop the package/receiver prefix, keep just the function name
			}
			return BucketKey{Filepath: frame.File, Lineno: frame.Line, Function: fn}
		}
		if !more {
			break
		}
	}
	return BucketKey{Filepath: "?", Lineno: 0, Function: "?"}
}

// CallInfo carries everything a caller knows about one call beyond its tokens.
//
// A struct rather than more parameters because the set has grown twice — cache
// counts, then latency — and every addition to a positional signature breaks
// every existing caller. Zero values mean "not reported", which is the honest
// default for all three.
type CallInfo struct {
	// inputTokens must be the *uncached* prompt when these are set. That is the
	// Anthropic convention, where a cache read is reported beside the prompt
	// rather than inside it, and it is what the ingest endpoints expect.
	// OpenAI-shaped usage is the other way round — prompt_tokens already
	// includes cached_tokens — so subtract the cached count out before calling.
	// An inclusive count bills the cached tokens twice, once at full rate and
	// once at the cache rate, and the platform freezes cost at receive.
	CacheReadTokens     int
	CacheCreationTokens int

	// LatencyMs is how long the call took. This SDK cannot measure it: it is
	// called after your LLM call returns, not around it, so there is no span to
	// time. Yours to measure, ours to record. Zero stores as 0, which the
	// Workspace reads as "not measured" rather than as an impossibly fast call.
	LatencyMs float64

	// TraceID ties this spend to the business events emitted during the same
	// run, so the backend matches them explicitly instead of falling back to a
	// time window. Empty means "no run id", which is the correct default for a
	// caller that never had one.
	TraceID string
}

// TracedLLMCall records an LLM call, attributed to the caller's source line, so
// the platform can price it. Nothing is returned, deliberately: the price table
// lives in exactly one place — the backend, which recomputes cost from the model and the
// four token counts at receive time. This SDK ships no price table. A bundled
// one is a second copy that drifts from the server's, and once it did, a row
// naming a model the SDK knew and the server didn't stopped being repriced.
func TracedLLMCall(model string, inputTokens, outputTokens int) {
	TracedLLMCallInfo(model, inputTokens, outputTokens, CallInfo{})
}

// TracedLLMCallInfo is TracedLLMCall for a call that reported cache tokens,
// a duration, a trace id, or any combination of the three.
func TracedLLMCallInfo(model string, inputTokens, outputTokens int, info CallInfo) {
	site := resolveCallerFrame()
	// Grab the source around the line before recording, so Push can send it
	// alongside the chain frame. Best-effort — a miss changes nothing here.
	captureSnippet(site.Filepath, site.Lineno)
	key := site.Site()
	key.Model = model
	CostProcessor.record(
		key, inputTokens, outputTokens, info.CacheReadTokens, info.CacheCreationTokens,
		info.LatencyMs, info.TraceID)
}

// ValueByFile tracks revenue/value earned, attributed to the calling file —
// the same "agent" identity TracedLLMCall uses for spend.
var ValueByFile = struct {
	mu     sync.Mutex
	values map[string]float64
}{values: make(map[string]float64)}

// ReportValue records value this agent earned (a closed deal, a resolved
// ticket, whatever the caller's business counts). Pushed home via PushValue.
func ReportValue(valueUSD float64) {
	key := resolveCallerFrame()
	ValueByFile.mu.Lock()
	defer ValueByFile.mu.Unlock()
	ValueByFile.values[key.Filepath] += valueUSD
}

// ValueSnapshot returns a copy of the current filepath -> value_usd totals.
func ValueSnapshot() map[string]float64 {
	ValueByFile.mu.Lock()
	defer ValueByFile.mu.Unlock()
	out := make(map[string]float64, len(ValueByFile.values))
	for k, v := range ValueByFile.values {
		out[k] = v
	}
	return out
}

// ClearValue resets recorded value — mainly useful between test runs.
func ClearValue() {
	ValueByFile.mu.Lock()
	defer ValueByFile.mu.Unlock()
	ValueByFile.values = make(map[string]float64)
}

File: node/src/tracing.js
// Attributes a (simulated) LLM call to the user-code line that issued it,
// walking the call stack past this file and past anything in node_modules
// (a reasonable proxy for "framework code" in Node — your own code is
// essentially never installed there). Mirrors python/aight/tracing.py.
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'

const THIS_FILE = fileURLToPath(import.meta.url)

// V8 stack frame lines look like:
//   "    at functionName (/path/to/file.js:10:15)"
//   "    at /path/to/file.js:10:15"                 (anonymous)
const FRAME_RE = /^\s*at\s+(?:(.+?)\s+\()?(.+?):(\d+):(\d+)\)?$/

function normalizePath(raw) {
  // ESM stack frames report file:// URLs; CJS ones report plain paths.
  return raw.startsWith('file://') ? fileURLToPath(raw) : raw
}

function resolveCallerFrame() {
  const holder = {}
  Error.captureStackTrace(holder, resolveCallerFrame)
  const lines = holder.stack.split('\n').slice(1) // drop the "Error" header line
  for (const line of lines) {
    const m = FRAME_RE.exec(line)
    if (!m) continue
    const [, fn, rawFilepath, lineno] = m
    const filepath = normalizePath(rawFilepath)
    if (filepath === THIS_FILE) continue
    if (filepath.includes('/node_modules/')) continue
    return { filepath, lineno: Number(lineno), function: fn || '<anonymous>' }
  }
  return { filepath: '?', lineno: 0, function: '?' }
}

// How many source lines to grab on each side of an attributed line, for the
// Line review code panel. Not configurable — a fixed small window is plenty for
// "what's on this line" context and keeps payloads small.
const SNIPPET_CONTEXT = 3

// "filepath\x00lineno" -> {start, text}: the first line number in the window and
// the joined source. Populated once per unique line ever seen, read by
// push() to attach source context to each chain frame. Mirrors tracing.py's
// SNIPPETS.
const SNIPPETS = new Map()

function snippetKey(filepath, lineno) {
  return `${filepath}\x00${lineno}`
}

/**
 * Best-effort: if the file isn't readable (deployed without source, a path from
 * a different machine, a stdin frame) just skip it — a missing snippet degrades
 * to the existing "not captured" UI state, and is never worth throwing over.
 * Already-seen lines are never re-read.
 */
function captureSnippet(filepath, lineno) {
  const key = snippetKey(filepath, lineno)
  if (SNIPPETS.has(key)) return
  try {
    const lines = readFileSync(filepath, 'utf8').split(/\r?\n/)
    const start = Math.max(1, lineno - SNIPPET_CONTEXT)
    const window = lines.slice(start - 1, lineno + SNIPPET_CONTEXT)
    if (window.length) {
      SNIPPETS.set(key, { start, text: window.join('\n').replace(/\n+$/, '') })
    }
  } catch {
    // Unreadable file — no snippet, no exception.
  }
}

/** The captured source window for a line, or undefined if there isn't one. */
export function snippetFor(filepath, lineno) {
  return SNIPPETS.get(snippetKey(filepath, lineno))
}

// GenAI semantic conventions have used both key names across versions; check
// both rather than assuming one. Confirm which your installed
// @traceloop/instrumentation-* version actually emits before relying on this.
//
// The third element is a convention, not a name, and the two disagree about
// whether the cached tokens are already inside the input count:
//
//   ...input_tokens   Anthropic-shaped — the uncached prompt only, with the
//                     cache counts reported separately alongside it.
//   ...prompt_tokens  OpenAI-shaped — the *total* prompt, cached tokens
//                     included.
//
// The wire wants the exclusive one, so an inclusive match subtracts the cache
// read back out below. Skipping that bills a cached token twice — once at the
// full input rate, again at the cache rate — and because the platform freezes
// cost at receive the overcount is permanent. Mirrors tracing.py.
const TOKEN_ATTR_CANDIDATES = [
  ['gen_ai.usage.input_tokens', 'gen_ai.usage.output_tokens', false],
  ['gen_ai.usage.prompt_tokens', 'gen_ai.usage.completion_tokens', true],
]

// Where the cache counts live, when the instrumentation reports them at all.
// Both spellings of the read are Anthropic's and the semconv has not settled on
// one; a missing key means "no caching reported", which is also the normal case
// for a provider that has none or a call that did not use it.
const CACHE_READ_ATTRS = ['gen_ai.usage.cache_read_input_tokens', 'gen_ai.usage.cached_tokens']
const CACHE_CREATION_ATTRS = ['gen_ai.usage.cache_creation_input_tokens']

function firstInt(attrs, keys) {
  for (const key of keys) {
    if (key in attrs) return Number(attrs[key] || 0)
  }
  return 0
}

/**
 * [input, output, cacheRead, cacheCreation] as the wire wants them, or null
 * when the span reported no usage at all. `input` is always the *uncached*
 * prompt by the time this returns — the one guarantee the function exists to
 * make.
 */
function usageFrom(attrs) {
  for (const [inputKey, outputKey, inclusive] of TOKEN_ATTR_CANDIDATES) {
    if (!(inputKey in attrs) || !(outputKey in attrs)) continue
    const cacheRead = firstInt(attrs, CACHE_READ_ATTRS)
    const cacheCreation = firstInt(attrs, CACHE_CREATION_ATTRS)
    const prompt = Number(attrs[inputKey] || 0)
    // Only an inclusive count can contain a cache figure, and only the read
    // one: a cache *write* is tokens being processed for the first time, not a
    // repeat of ones already in the prompt. Clamped at zero because
    // instrumentation reporting a cache count larger than the prompt it is
    // supposedly inside is telling us its own numbers disagree, and a negative
    // token count would be worse than a floor.
    const uncached = inclusive ? Math.max(prompt - cacheRead, 0) : prompt
    return [uncached, Number(attrs[outputKey] || 0), cacheRead, cacheCreation]
  }
  return null
}

class Bucket {
  constructor() {
    this.calls = 0
    this.inputTokens = 0
    this.outputTokens = 0
    // Kept beside inputTokens, never folded into it. The platform sums four
    // rates and freezes the result at receive, so an input count that already
    // contained the cached tokens would bill them twice — once at the full
    // input rate and again at the cache rate — and the overcount could never
    // be corrected.
    this.cacheReadTokens = 0
    this.cacheCreationTokens = 0
    // Summed, not averaged: the bucket divides by calls when it reports.
    // Only ever what the caller measured and passed — this SDK never times a
    // call itself, and remote.js omits the key when nobody did.
    this.latencyMs = 0
    // Last non-empty wins, matching Python: a line that recorded once with a
    // trace id and once without is still the traced run. '' means nobody set
    // one, which the server reads as "no explicit match available".
    this.traceId = ''
  }
}

export class CostByLineProcessor {
  constructor() {
    this.buckets = new Map() // "filepath\x00lineno\x00function\x00model" -> Bucket
  }

  /**
   * Records one call. The model is part of the bucket key, not just a field on
   * the bucket: a line that calls two models stays two buckets, each priced by
   * the server on its own tokens. Merging them summed both models' tokens under
   * whichever name was recorded last, and the server then repriced every cheap
   * call at the expensive model's rate.
   */
  record(
    filepath,
    lineno,
    fn,
    model,
    inputTokens,
    outputTokens,
    cacheReadTokens = 0,
    cacheCreationTokens = 0,
    latencyMs = 0,
    traceId = ''
  ) {
    const key = `${filepath}\x00${lineno}\x00${fn}\x00${model}`
    let bucket = this.buckets.get(key)
    if (!bucket) {
      bucket = new Bucket()
      this.buckets.set(key, bucket)
    }
    bucket.calls += 1
    bucket.inputTokens += inputTokens
    bucket.outputTokens += outputTokens
    bucket.cacheReadTokens += cacheReadTokens
    bucket.cacheCreationTokens += cacheCreationTokens
    bucket.latencyMs += latencyMs
    if (traceId) bucket.traceId = traceId
  }

  /**
   * One entry per (call site, model): {filepath, lineno, function, model,
   * calls, inputTokens, outputTokens, cacheReadTokens, cacheCreationTokens,
   * latencyMs, traceId}. Nothing is priced here.
   */
  entries() {
    return [...this.buckets.entries()].map(([key, bucket]) => {
      const [filepath, lineno, fn, model] = key.split('\x00')
      return { filepath, lineno: Number(lineno), function: fn, model, ...bucket }
    })
  }

  /** Drops everything recorded so far. push() calls this once the server has
   * acknowledged, so a later push sends only what happened since — see the
   * double-counting note in remote.js. Mirrors Go's Clear() and Java's clear(). */
  clear() {
    this.buckets.clear()
  }

  /**
   * The OTel SpanProcessor surface, so autoInstrument() can register this same
   * processor on the TracerProvider it installs: a span an auto-instrumentor
   * emits lands in the buckets a hand-wrapped tracedLlmCall writes to. Mirrors
   * CostByLineProcessor.on_start / on_end in tracing.py.
   *
   * onStart captures the caller's line while the instrumented call is still on
   * the stack; onEnd reads it back with the model and usage the instrumentor
   * attached. A span carrying neither is not a priced call (an HTTP client's, a
   * DB driver's) and is skipped — this processor sees every span in the host
   * process, and recording the rest would fill the Workspace with calls=1 rows
   * that name no model and no tokens.
   */
  onStart(span) {
    // No "already attributed" guard needed: resolveCallerFrame() walks past
    // node_modules, so an instrumentor's own nested spans (an HTTP client's
    // inside the provider's) resolve to the same user line as the outer one.
    const { filepath, lineno, function: fn } = resolveCallerFrame()
    span.setAttribute('code.chain.length', 1)
    span.setAttribute('code.chain.0.filepath', filepath)
    span.setAttribute('code.chain.0.lineno', lineno)
    span.setAttribute('code.chain.0.function', fn)
  }

  onEnd(span) {
    const attrs = span.attributes || {}
    const model = attrs['gen_ai.request.model'] || ''
    const usage = usageFrom(attrs)
    if (!model || !usage) return
    const filepath = attrs['code.chain.0.filepath'] || '?'
    const lineno = Number(attrs['code.chain.0.lineno'] || 0)
    const fn = attrs['code.chain.0.function'] || '?'
    captureSnippet(filepath, lineno)
    this.record(filepath, lineno, fn, model, ...usage, 0, attrs['aight.trace_id'] || '')
  }

  shutdown() {}

  forceFlush() {
    return Promise.resolve()
  }

  /**
   * Plain-text report of what was traced locally, calls first. There is no $
   * figure by design: this SDK carries no price table, so the only honest
   * thing it can show is the call count and the tokens the server will price.
   */
  report() {
    return this.entries()
      .sort((a, b) => b.calls - a.calls)
      .map((e) => {
        const base = e.filepath.split('/').pop()
        const model = e.model ? ` [${e.model}]` : ''
        return `${base}:${e.lineno} (${e.function})${model} = ${e.calls} call(s), ${e.inputTokens} in / ${e.outputTokens} out`
      })
      .join('\n')
  }
}

export const COST_PROCESSOR = new CostByLineProcessor()

/**
 * Record a (simulated) LLM call, attributed to the caller's source line, so the
 * platform can price it. Returns nothing: this SDK carries no price table by
 * design — the server prices every row from its own, and a bundled copy here
 * would be a second table that drifts from it.
 *
 * Pass `cacheReadTokens` / `cacheCreationTokens` for a call that used prompt
 * caching — omitting them under-reports badly, since a cache read is about a
 * tenth of the input rate and a write about 125%.
 *
 * `inputTokens` must be the *uncached* prompt. That is the Anthropic
 * convention, where a cache read is reported beside the prompt rather than
 * inside it, and it is what the ingest endpoints expect. OpenAI-shaped usage is
 * the other way round — `prompt_tokens` already contains `cached_tokens` — so
 * subtract the cached count out before calling. An inclusive count here bills
 * the cached tokens twice, and the platform freezes cost at receive, so the
 * overcount is permanent.
 *
 * `latencyMs` is yours to measure. This SDK does not time the LLM call: it is
 * called after your call returns rather than around it, so there is nothing
 * here to time. Time the call yourself and pass the milliseconds; omit it and
 * no latency is claimed, and the Workspace shows "—" on the Speed axis rather
 * than an impossibly fast number.
 *
 * `traceId` is what upgrades a later pushEvent() from the server's time-window
 * guess to an EXPLICIT match: record your agent run's id here and report the
 * same id on the business event that run produced, and the two are tied
 * together without any inference. Omitting it stores '' — no match, not a wrong
 * one.
 *
 * The source line this attributes to is read once and cached, and its
 * surrounding lines ride along with the next push() as `snippet`. Unreadable
 * source is silently skipped.
 */
export function tracedLlmCall(
  model,
  inputTokens,
  outputTokens,
  cacheReadTokens = 0,
  cacheCreationTokens = 0,
  latencyMs = 0,
  traceId = ''
) {
  const { filepath, lineno, function: fn } = resolveCallerFrame()
  captureSnippet(filepath, lineno)
  COST_PROCESSOR.record(
    filepath,
    lineno,
    fn,
    model,
    inputTokens,
    outputTokens,
    cacheReadTokens,
    cacheCreationTokens,
    latencyMs,
    traceId
  )
}

/** filepath -> total value_usd earned. Reported via reportValue(), pushed via pushValue(). */
export const VALUE_BY_FILE = new Map()

/** Record revenue/value this agent earned, attributed to the calling file (the same "agent" identity tracedLlmCall uses for spend). */
export function reportValue(valueUsd) {
  const { filepath } = resolveCallerFrame()
  VALUE_BY_FILE.set(filepath, (VALUE_BY_FILE.get(filepath) || 0) + valueUsd)
}
